### PR TITLE
Performance boost

### PR DESCRIPTION
**Pull request by juliushaertl**
Improvements are split into individual commits for easier review. 

Overall this adds additional caching layers and groups queries where we performed them for n results into one.

## ToDo

- [x] Get in #3793 first / might fix OverviewService to perform queries per board
- [x] StackMapper::find could use some caching on request level
- [x] BoardService::find does not need full details (users/settings/sessions) when we set the relative board on cards 
- [x] Group queries for related stacks
- [x] Make sure that in queries do not exceed the oracle limit: More than 1000 expressions in a list are not allowed on Oracle.

## Follow up issues
- [ ] CardMapper findBoardId cache should probably use something like WithLocalCache to avoid heavy round trips to redis (which happens at least once for every file share)
- [ ] Enrichment of board data could be grouped queries (might only have a noticeable effect on hundreds of boards a user has)

### Board list 

- Removes 1 query per board the user has access to
- Removes 1 additional query

Screenshot 2023-02-17 at 09 50 43

### Overview

- Removes 2-4 queries per card that is listed in the overview (extra queries still needed if a card has comments to fetch the unread counter)

Screenshot 2023-02-17 at 09 51 52

### Single board request

- removes 1 query for each sharing rule that a board has

Screenshot 2023-02-17 at 09 53 45

### Stack list

- removes 2-4 queries per card that is on the board
- reduce query to fetch stack details from the amount of cards to the number of stacks (40 cards, 4 stacks -> 40 to 4 queries)

Screenshot 2023-02-17 at 09 55 15

### Propfind on Deck folder (3200 incoming deck file shares)

Screenshot 2023-02-17 at 09 59 12

## Testing

<details>
<summary>Script to generate random board content</summary>

```
<?php
/*
 * @copyright Copyright (c) 2021 Julius Härtl <jus@bitgrid.net>
 *
 * @author Julius Härtl <jus@bitgrid.net>
 *
 * @license GNU AGPL version 3 or any later version
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */

declare(strict_types=1);

if ($argc === 1) {
	echo 'Usage: php deck-random.php [userId] [boardCount] [stackCount] [cardCount]';
	exit;
}

require_once __DIR__ . '/../lib/versioncheck.php';

function generateRandomString($length = 10) {
	$characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
	$charactersLength = strlen($characters);
	$randomString = '';
	for ($i = 0; $i < $length; $i++) {
		$randomString .= $characters[random_int(0, $charactersLength - 1)];
	}
	return $randomString;
}

function random_color_part() {
	return str_pad(dechex(mt_rand(0, 255)), 2, '0', STR_PAD_LEFT);
}

function random_color() {
	return random_color_part() . random_color_part() . random_color_part();
}

function randomDateInRange(DateTime $start, DateTime $end) {
	$randomTimestamp = mt_rand($start->getTimestamp(), $end->getTimestamp());
	$randomDate = new DateTime();
	$randomDate->setTimestamp($randomTimestamp);
	return $randomDate;
}

try {
	require_once __DIR__ . '/../lib/base.php';
	OC_App::loadApps(['authentication']);
	OC_App::loadApps(['filesystem', 'logging']);
	OC_App::loadApps();

	$user = \OC::$server->getUserManager()->get($argv[1] ?? 'admin');
	$uid = $user->getUID() ?? 'admin';
	\OC::$server->getUserSession()->setUser($user);

	$boardCount = $argv[2] ?? 1;
	$stackCount = $argv[3] ?? 4;
	$cardCount = $argv[4] ?? 10;
	$attachmentCount = 10;

	/** @var \OCA\Deck\Service\BoardService $boardService */
	$boardService = \OC::$server->get(\OCA\Deck\Service\BoardService::class);
	/** @var \OCA\Deck\Service\StackService $stackService */
	$stackService = \OC::$server->get(\OCA\Deck\Service\StackService::class);
	/** @var \OCA\Deck\Service\CardService $cardService */
	$cardService = \OC::$server->get(\OCA\Deck\Service\CardService::class);

	$shareManager = \OCP\Server::get(\OCP\Share\IManager::class);

	$root = \OCP\Server::get(\OCP\Files\IRootFolder::class)->getUserFolder($uid);
	$deckFolder = $root->nodeExists('deck_random') ? $root->get('deck_random') : $root->newFolder('deck_random');
	for ($h = 0; $h < $boardCount; $h++) {
		$board = $boardService->create(generateRandomString(), $uid, random_color());
		$users = \OC::$server->getUserManager()->search('');
		$acls = array_rand($users, 5);
		$boardService->addAcl($board->getId(), \OCA\Deck\Db\Acl::PERMISSION_TYPE_USER, 'bob', true, true, true);
		foreach ($acls as $acl) {
			if ($acl !== $uid && $acl !== 'bob') {
				$boardService->addAcl($board->getId(), \OCA\Deck\Db\Acl::PERMISSION_TYPE_USER, $acl, true, true, true);
			}
		}
		for ($i = 0; $i < $stackCount; $i++) {
			echo "Created stack $i\n";
			$stack = $stackService->create(generateRandomString(), $board->getId(), -1);
			for ($j = 0; $j < $cardCount; $j++) {
				echo "Created card $i.$j\n";
				$randomDate = date("Y-m-d H:i:s", mt_rand(1262055681, 1262055681));
				$card = $cardService->create(generateRandomString(100), $stack->getId(), 'text', -1, $uid, generateRandomString(2000), null);

				for ($ai = 0; $ai < $attachmentCount; $ai++) {
					$file = $deckFolder->newFile('attachment-' . $card->getId() . '-' . $ai . '.txt', $randomDate);
					$share = $shareManager->newShare();
					$share->setSharedBy($uid);
					$share->setSharedWith((string)$card->getId());
					$share->setPermissions(1);
					$share->setShareType(\OCP\Share\IShare::TYPE_DECK);
					$share->setNode($file);
					$shareManager->createShare($share);
				}
			}
		}
	}
} catch (\Throwable $e) {
	echo $e->getMessage() . PHP_EOL;
	echo $e->getTraceAsString() . PHP_EOL;
}
```

</details>

### CODE REVIEW

File: .github/workflows/integration.yml [hunk 1]
```diff
@@ -91,11 +92,63 @@ jobs:
...
+
+            myOutput = ''
+            await exec.exec('cat', ['/home/runner/work/${{ env.APP_NAME }}/${{ env.APP_NAME }}/apps/${{ env.APP_NAME }}/tests/integration/base-query-count.txt'], options)
+            const baseCount = parseInt(myOutput, 10)
+
+            const absoluteIncrease = queryCount - baseCount
+            const relativeIncrease = baseCount <= 0 ? 100 : (parseInt((absoluteIncrease / baseCount * 10000), 10) / 100)
+
+            if (absoluteIncrease >= 100 || relativeIncrease > 5) {
+              const comment = `🐢 Performance warning.\nIt looks like the query count of the integration tests increased with this PR.\nDatabase query count is now ` + queryCount + ' was ' + baseCount + ' (+' + relativeIncrease + '%)\nPlease check your code again. If you added a new test this can be expected and the base value in tests/integration/base-query-count.txt can be increased.'
```
Comment (marcelklehr) [inline .github/workflows/integration.yml:138]: Wow. Love it :heart_eyes:

File: lib/Db/BoardMapper.php [hunk 2]
```diff
@@ -259,11 +270,7 @@ public function findAllByUser(string $userId, ?int $limit = null, ?int $offset =
 			$entry->setShared(1);
 		}
 		$entries = array_merge($entries, $sharedEntries);
-		/* @var Board $entry */
-		foreach ($entries as $entry) {
-			$acl = $this->aclMapper->findAll($entry->id);
-			$entry->setAcl($acl);
-		}
```
Comment (marcelklehr) [inline lib/Db/BoardMapper.php:266]: Sweet! Lot's of queries shaved off :)

File: lib/Db/User.php [hunk 3]
```diff
@@ -23,27 +23,30 @@
 
 namespace OCA\Deck\Db;
 
-use OCP\IUser;
+use OCP\IUserManager;
 
 class User extends RelationalObject {
```
Comment (marcelklehr) [inline lib/Db/User.php:28]: I'd love it if we could get that RelationalObject goodness into server at some point, seems very useful!

File: lib/Service/BoardService.php [hunk 4]
```diff
@@ -147,96 +148,45 @@ public function getUserBoards(?int $since = null, bool $includeArchived = true,
...
-			'PERMISSION_READ' => $permissions[Acl::PERMISSION_READ] ?? false,
-			'PERMISSION_EDIT' => $permissions[Acl::PERMISSION_EDIT] ?? false,
-			'PERMISSION_MANAGE' => $permissions[Acl::PERMISSION_MANAGE] ?? false,
-			'PERMISSION_SHARE' => $permissions[Acl::PERMISSION_SHARE] ?? false
-		]);
-		$this->enrichWithUsers($board);
-		$this->enrichWithBoardSettings($board);
-		$this->enrichWithActiveSessions($board);
-		$this->boardsCache[$board->getId()] = $board;
+		[$board] = $this->enrichBoards([$board], $fullDetails);
```
Comment (marcelklehr) [inline lib/Service/BoardService.php:186]: Nice refactoring!

File: .github/workflows/integration.yml [hunk 5]
```diff
@@ -91,11 +92,63 @@ jobs:
...
+            await exec.exec(`/bin/bash -c "cat /home/runner/work/${{ env.APP_NAME }}/${{ env.APP_NAME }}/query.log | wc -l"`, [], options)
+            msg = myOutput
+            const queryCount = parseInt(myOutput, 10)
+
+            myOutput = ''
+            await exec.exec('cat', ['/home/runner/work/${{ env.APP_NAME }}/${{ env.APP_NAME }}/apps/${{ env.APP_NAME }}/tests/integration/base-query-count.txt'], options)
+            const baseCount = parseInt(myOutput, 10)
+
+            const absoluteIncrease = queryCount - baseCount
+            const relativeIncrease = baseCount <= 0 ? 100 : (parseInt((absoluteIncrease / baseCount * 10000), 10) / 100)
```
Comment (marcelklehr) [inline .github/workflows/integration.yml:135]: Can we print these stats to stdout just for curiosity?

**Review by marcelklehr — APPROVED** [review]

**Comment by marcelklehr** [comment]
How much impact do these changes have? out of curiosity?

File: .github/workflows/integration.yml (hunk 1 above)
Comment (juliushaertl) [reply .github/workflows/integration.yml:138]: All credits to Joas for it https://github.com/nextcloud/notifications/blob/master/.github/workflows/performance.yml 😉

**Comment by juliushaertl** [comment]
Added summary of the saved queries to the first post above each blackfire screenshot ;)

File: lib/Db/User.php (hunk 3 above)
Comment (juliushaertl) [reply lib/Db/User.php:28]: I'm not too happy anymore with the approach and think it should actually be decoupled from the entity itself. I'm more in favour of having separate Model classes in the future like Talk has.

**Pull request merged**